### PR TITLE
Dynamic port 

### LINKED ISSUE #773: CassandraSinkCluster dynamic port handling
Currently we hardcode the port 9042 everywhere, but we should handle the port dynamically from the config and system.peers_v2 results.

### PR DESCRIPTION
**Pull request by conorbros**
closes #773

### CODE REVIEW

**Comment by conorbros** [comment]
Ready for some reviews, still need to add some tests and documentation about using non-standard ports.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 1]
```diff
@@ -103,68 +70,281 @@ async fn topology_task_process(
...
+                            let port = if let Some(broadcast_port_index) = broadcast_port_index {
+                                if let Some(MessageValue::Integer(value, _)) =
+                                    row.get(broadcast_port_index)
+                                {
+                                    *value
+                                } else {
+                                    return Err(anyhow!("port not an integer"));
+                                }
+                            } else {
+                                9042
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:188]: can we add a comment just above the 9042, like: "cassandra 3.x does not have the broadcast_port field so we just assume the default port.

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 2]
```diff
@@ -103,68 +70,281 @@ async fn topology_task_process(
...
+                            } else {
+                                false
+                            }
+                        })
+                        .map(|row| {
+                            let tokens =
+                                if let Some(MessageValue::List(list)) = row.get(tokens_index) {
+                                    list.iter()
+                                        .map::<Result<String>, _>(|x| match x {
+                                            MessageValue::Varchar(a) => Ok(a.clone()),
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:156]: You can avoid cloning all of these strings like this:
image

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 3]
```diff
@@ -103,68 +70,281 @@ async fn topology_task_process(
...
+                                value
+                            } else {
+                                return Err(anyhow!("port is not an integer"));
+                            }
+                        } else {
+                            9042
+                        };
+
+                        Ok(CassandraNode {
+                            address: SocketAddr::new(ip, port.try_into()?),
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:331]: useless `.try_into()?`?

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 4]
```diff
@@ -103,68 +70,281 @@ async fn topology_task_process(
...
+                        };
+
+                        let port = if let Some(message_value) = row.pop() {
+                            if let MessageValue::Integer(value, _) = message_value {
+                                value
+                            } else {
+                                return Err(anyhow!("port is not an integer"));
+                            }
+                        } else {
+                            9042
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:327]: same here but comment about system.peers_v2 vs system.peers

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs [hunk 5]
```diff
@@ -103,68 +70,281 @@ async fn topology_task_process(
...
+                })),
+                tx,
+            )?;
+            response = rx.await?.response?;
+        }
+
+        into_nodes(response, data_center)
+    }
+
+    fn is_unconfigured_peers_v2_table_response(message: &mut Message) -> bool {
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:263]: How about: `is_peers_v2_does_not_exist_error`.
That way you dont need to know that "unconfigured" is what cassandra says when a table does not exist.

**Review by rukai — COMMENTED** [review]
General approach looks good, heres some minor feedback

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs (hunk 3 above)
Comment (conorbros) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:331]: image

Am I missing something?

Comment (rukai) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:333]: No, I was, my bad.

:+1:

File: shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs (hunk 4 above)
Comment (rukai) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:329]: comment doesnt quite make sense directly copied.
How about something like: "this method supports both system.peers and system.peers_v2, system.peers does not have a field for the port so we fallback to the default port."

Comment (conorbros) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/topology.rs:329]: Ah yep, MB.

**Review by benbromhead — APPROVED** [review]

File: shotover-proxy/tests/helpers/config.yaml [hunk 6]
```diff
@@ -1,3 +1,3 @@
 ---
-main_log_level: "warn,shotover_proxy=warn"
+main_log_level: "info,shotover_proxy=info"
```
Comment (rukai) [inline shotover-proxy/tests/helpers/config.yaml:2]: ^

File: shotover-proxy/src/transforms/cassandra/sink_cluster/mod.rs [hunk 7]
```diff
@@ -638,6 +640,8 @@ impl CassandraSinkCluster {
                         listen_address_alias = alias;
                     } else if column.name == Identifier::Unquoted("host_id".to_string()) {
                         host_id_alias = alias;
+                    } else if column.name == Identifier::Unquoted("rpc_port".to_string()) {
```
Comment (rukai) [inline shotover-proxy/src/transforms/cassandra/sink_cluster/mod.rs:643]: I dont believe such a field exists:
image

Comment (conorbros) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/mod.rs:643]: Oops got mixed up with `system.local` clearly.

Comment (conorbros) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/mod.rs:643]: Oh wait I was correct, the function that change is in is `rewrite_table_local`.

Comment (rukai) [reply shotover-proxy/src/transforms/cassandra/sink_cluster/mod.rs:643]: Oh true, I got the tables mixed up.
rpc_address/rpc_port is actually the fields describing where the cql protocol connection goes. :+1:

**Review by rukai — APPROVED** [review]
very nice!

**Pull request merged**